File: packages/typescript/examples/12_file_handling.ts
/**
 * Multi-modal file handling.
 *
 * Run: npx ts-node examples/12_file_handling.ts
 */
import { Agent, FilePart, Artifact } from '../src';

const agent = new Agent({ name: 'FileProcessor', description: 'Processes files' });

agent.skill('summarize_doc', { description: 'Summarize a document' }, async ({ document }: { document: { name: string; data: string; mimeType?: string } }) => {
  const file = new FilePart({ name: document.name, data: Buffer.from(document.data, 'base64'), mimeType: document.mimeType });
  const content = file.data?.toString('utf-8') || '';
  const words = content.split(/\s+/);
  const summary = words.slice(0, 50).join(' ') + '...';
  return 'Summary of ' + file.name + ': ' + summary;
});

agent.skill('process_data', { description: 'Process JSON data' }, async ({ data }: { data: { items?: Array<{ value?: number }> } }) => {
  const items = data.items || [];
  return { processed: items.length, total: items.reduce((sum, item) => sum + (item.value || 0), 0) };
});

agent.skill('generate_report', { description: 'Generate a report' }, async ({ title }: { title: string }) => {
  const artifact = new Artifact({ name: 'report.json', description: 'Report: ' + title });
  artifact.addText('# Report: ' + title + '\n\nThis is the summary...');
  artifact.addData({ title, generated: new Date().toISOString().split('T')[0], metrics: { users: 100, revenue: 5000 } });
  return artifact.toA2A();
});

agent.skill('analyze_image', { description: 'Analyze an image' }, async ({ image }: { image: { name: string; data: string; mimeType: string } }) => {
  if (!image.mimeType.startsWith('image/')) return { error: 'Expected image, got ' + image.mimeType };
  const data = Buffer.from(image.data, 'base64');
  return { filename: image.name, mime_type: image.mimeType, size_bytes: data.length };
});

agent.run({ port: 8787 });

File: packages/java/examples/12_FileHandling.java
import com.a2alite.Agent;
import com.a2alite.SkillConfig;
import java.nio.charset.StandardCharsets;
import java.time.LocalDate;
import java.util.*;

/**
 * Multi-modal file handling.
 *
 * Run: ./gradlew run -PmainClass=FileHandling
 */
class FileHandling {
    public static void main(String[] args) {
        var agent = Agent.builder().name("FileProcessor").description("Processes files").build();

        agent.skill("summarize_doc", SkillConfig.of("Summarize a document"), params -> {
            @SuppressWarnings("unchecked")
            Map<String, Object> doc = (Map<String, Object>) params.get("document");
            String name = (String) doc.get("name");
            byte[] decoded = Base64.getDecoder().decode((String) doc.get("data"));
            String content = new String(decoded, StandardCharsets.UTF_8);
            String[] words = content.split("\\s+");
            String summary = String.join(" ", Arrays.copyOf(words, Math.min(50, words.length))) + "...";
            return "Summary of " + name + ": " + summary;
        });

        agent.skill("process_data", SkillConfig.of("Process JSON data"), params -> {
            @SuppressWarnings("unchecked")
            Map<String, Object> data = (Map<String, Object>) params.get("data");
            @SuppressWarnings("unchecked")
            List<Map<String, Object>> items = (List<Map<String, Object>>) data.getOrDefault("items", List.of());
            double total = items.stream().mapToDouble(i -> ((Number) i.getOrDefault("value", 0)).doubleValue()).sum();
            return Map.of("processed", items.size(), "total", total);
        });

        agent.skill("generate_report", SkillConfig.of("Generate a report"), params -> {
            String title = (String) params.get("title");
            return Map.of("name", "report.json", "parts", List.of(
                Map.of("kind", "text", "text", "# Report: " + title),
                Map.of("kind", "data", "data", Map.of("title", title, "generated", LocalDate.now().toString()))
            ));
        });

        agent.skill("analyze_image", SkillConfig.of("Analyze an image"), params -> {
            @SuppressWarnings("unchecked")
            Map<String, Object> image = (Map<String, Object>) params.get("image");
            String mimeType = (String) image.get("mimeType");
            if (!mimeType.startsWith("image/")) return Map.of("error", "Expected image, got " + mimeType);
            byte[] decoded = Base64.getDecoder().decode((String) image.get("data"));
            return Map.of("filename", image.get("name"), "mime_type", mimeType, "size_bytes", decoded.length);
        });

        agent.run(8787);
    }
}
